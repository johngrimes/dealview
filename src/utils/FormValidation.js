// @flow

type ValidatorFunction = (value: string) => boolean
export type FieldErrors = string[]

//
// Validates a set of fields, using a supplied validator function.
//
const validate = (value: string, validator: ValidatorFunction, message: string): FieldErrors => {
  if (!validator(value)) return [message]
  else return []
}

//
// Validates the presence of a set of fields.
//
export const required = (value: string): FieldErrors =>
  validate(value, v => !!v, "can't be blank")

//
// Validates that a set of fields have values with a minimum length.
//
export const minLength = (value: string, min: number): FieldErrors =>
  validate(value, v => !v || v.length >= min, `must be at least ${min} characters`)

//
// Validates that a set of fields have values with a maximum length.
//
export const maxLength = (value: string, max: number): FieldErrors =>
  validate(value, v => !v || v.length <= max, `must be at most ${max} characters`)

//
// Validates that a set of fields have values with an exact length.
//
export const isLength = (value: string, is: number): FieldErrors =>
  validate(value, v => !v || v.length === is, `must be exactly ${is} characters`)
