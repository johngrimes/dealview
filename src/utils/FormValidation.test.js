/* global expect */

import * as Validation from 'utils/FormValidation'

describe('areErrorsPresent', () => {
  it('should return false for empty FormErrors object', () => {
    const errors = {
      name: [],
      address: {
        line1: [],
        line2: [],
        line3: [],
      },
      notes: [],
    }
    expect(Validation.areErrorsPresent(errors)).toBe(false)
  })

  it('should return true for FormErrors with nested value', () => {
    const errors = {
      name: [],
      address: {
        line1: ['does not sound right'],
        line2: [],
        line3: [],
      },
      notes: [],
    }
    expect(Validation.areErrorsPresent(errors)).toBe(true)
  })
})

describe('required', () => {
  it('should return an error for a blank string', () => {
    expect(Validation.required('')).toMatchSnapshot()
  })

  it('should return an error for undefined', () => {
    expect(Validation.required(undefined)).toMatchSnapshot()
  })

  it('should not return an error for a string', () => {
    expect(Validation.required('dog')).toMatchSnapshot()
  })

  it('should not return an error for a number', () => {
    expect(Validation.required(54)).toMatchSnapshot()
  })
})

describe('minLength', () => {
  it('should return an error for a string that is less', () => {
    expect(Validation.minLength('dog', 4)).toMatchSnapshot()
  })

  it('should not return an error for a string that is equal', () => {
    expect(Validation.minLength('dog', 3)).toMatchSnapshot()
  })

  it('should not return an error for a string that is more', () => {
    expect(Validation.minLength('dog', 2)).toMatchSnapshot()
  })
})

describe('maxLength', () => {
  it('should return an error for a string that is more', () => {
    expect(Validation.maxLength('horse', 4)).toMatchSnapshot()
  })

  it('should not return an error for a string that is equal', () => {
    expect(Validation.maxLength('horse', 5)).toMatchSnapshot()
  })

  it('should not return an error for a string that is less', () => {
    expect(Validation.maxLength('horse', 6)).toMatchSnapshot()
  })
})

describe('isLength', () => {
  it('should return an error for a string that is more', () => {
    expect(Validation.isLength('horse', 4)).toMatchSnapshot()
  })

  it('should not return an error for a string that is equal', () => {
    expect(Validation.isLength('horse', 5)).toMatchSnapshot()
  })

  it('should return an error for a string that is less', () => {
    expect(Validation.isLength('horse', 6)).toMatchSnapshot()
  })
})
