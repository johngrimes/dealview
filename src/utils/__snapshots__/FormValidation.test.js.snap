// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isLength should not return an error for a string that is equal 1`] = `Array []`;

exports[`isLength should return an error for a string that is less 1`] = `
Array [
  "must be exactly 6 characters",
]
`;

exports[`isLength should return an error for a string that is more 1`] = `
Array [
  "must be exactly 4 characters",
]
`;

exports[`maxLength should not return an error for a string that is equal 1`] = `Array []`;

exports[`maxLength should not return an error for a string that is less 1`] = `Array []`;

exports[`maxLength should return an error for a string that is more 1`] = `
Array [
  "must be at most 4 characters",
]
`;

exports[`minLength should not return an error for a string that is equal 1`] = `Array []`;

exports[`minLength should not return an error for a string that is more 1`] = `Array []`;

exports[`minLength should return an error for a string that is less 1`] = `
Array [
  "must be at least 4 characters",
]
`;

exports[`required should not return an error for a number 1`] = `Array []`;

exports[`required should not return an error for a string 1`] = `Array []`;

exports[`required should return an error for a blank string 1`] = `
Array [
  "can't be blank",
]
`;

exports[`required should return an error for undefined 1`] = `
Array [
  "can't be blank",
]
`;
